src/db/api/repository.ts

```diff
@@ -0,0 +1,78 @@
+// tslint:disable-next-line: import-name
+import BN from 'bn.js';
+import { AnySSZType, deserialize, hashTreeRoot, serialize } from '@chainsafe/ssz';
+import { IBeaconConfig } from '@chainsafe/eth2.0-config';
+
+import { IDatabaseController } from '../controller';
+import { Bucket, encodeKey } from '../schema';
+
+export type Id = Buffer | string | number | BN;
+
+export abstract class Repository<T> {
+    protected config: IBeaconConfig;
+
+    protected db: IDatabaseController;
+
+    protected bucket: Bucket;
+
+    protected type: AnySSZType;
+
+    protected constructor(config: IBeaconConfig, db: IDatabaseController, bucket: Bucket, type: AnySSZType) {
+        this.config = config;
+        this.db = db;
+        this.bucket = bucket;
+        this.type = type;
+    }
+
+    public async get(id: Id): Promise<T | null> {
+        try {
+            const value = await this.db.get(encodeKey(this.bucket, id));
+            if (!value) return null;
+            return deserialize(value, this.type);
+        } catch (e) {
+            return null;
+        }
+    }
+
+    public async has(id: Id): Promise<boolean> {
+        return (await this.get(id)) !== null;
+    }
+
+    public async setUnderRoot(value: T): Promise<void> {
+        await this.set(hashTreeRoot(value, this.type), value);
+    }
+
+    public async set(id: Id, value: T): Promise<void> {
+        await this.db.put(encodeKey(this.bucket, id), serialize(value, this.type));
```

Comment:
could you extract serialize/deserialize/hashTreeRoot into separate class (something like "CGSerialization) so we can change serialization to google protobufs at some point.